### PR TITLE
FIX: Avoid copying source script when `plot_html_show_source_link` is False in plot directive

### PR DESCRIPTION
**Pull request by harupy**
## PR Summary

In the plot directive, if `plot_html_show_source_link` is `False`, it's not necessary to copy the source scripts (sorry if I'm missing something). This change slightly decreases disk usage.

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by tacaswell** [comment]
This seems reasonable and correct to me.

Could you possibly add a few more asserts to https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_sphinxext.py to make sure the source is and is not copied in the relevant cases? 

The rst we are running against is https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/tinypages/some_plots.rst

**Comment by tacaswell** [comment]
Thank you for working on this and welcome 😄 (sorry should have lead with that!).

**Comment by harupy** [comment]
@tacaswell Added a few asserts. Could you take another look?

File: lib/matplotlib/tests/test_sphinxext.py [hunk 1]
```diff
@@ -95,12 +95,23 @@ def plot_directive_file(num):
...
-
-
-def build_sphinx_html(source_dir, doctree_dir, html_dir):
+    # Make sure that source scripts are generated by default
+    assert "some_plots-1.py" in [p.name for p in html_dir.iterdir()]
+    # Make sure that source scripts are NOT generated when
+    # `plot_html_show_source_link` is set to False
+    html_dir = source_dir / '_build' / 'html_no_source'
+    # `plot_html_show_source_link=0` is equivalent to
+    # `plot_html_show_source_link=False`
```
Comment (harupy) [inline lib/matplotlib/tests/test_sphinxext.py:104]: Ref: https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-D

File: lib/matplotlib/tests/test_sphinxext.py [hunk 2]
```diff
@@ -95,12 +95,23 @@ def plot_directive_file(num):
...
+    build_sphinx_html(source_dir, doctree_dir, html_dir,
+                      extra_args=["-D", "plot_html_show_source_link=0"])
+    assert "some_plots-1.py" not in [p.name for p in html_dir.iterdir()]
```
Comment (harupy) [inline lib/matplotlib/tests/test_sphinxext.py:107]: Should I make this more strict?

**Comment by harupy** [comment]
> I think it would be better to add another plot to the example file?

`plot_html_show_source_link` is a global configuration, not a directive-level configuration, so I don't think we can do something like:

```
.. plot::
     :show_source_link: False
     
     # code
```

**Review by tacaswell — APPROVED** [review]

File: lib/matplotlib/tests/test_sphinxext.py [hunk 3]
```diff
@@ -97,10 +97,33 @@ def plot_directive_file(num):
...
+                      extra_args=["-D", "plot_html_show_source_link=0"])
+    assert "index-1.py" not in [p.name for p in html_dir2.iterdir()]
+
+
+def build_sphinx_html(source_dir, doctree_dir, html_dir, extra_args=None):
     # Build the pages with warnings turned into errors
+    extra_args = [] if extra_args is None else extra_args
     cmd = [sys.executable, '-msphinx', '-W', '-b', 'html',
-           '-d', str(doctree_dir), str(source_dir), str(html_dir)]
+           '-d', str(doctree_dir), str(source_dir), str(html_dir)] + extra_args
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_sphinxext.py:126]:
```suggestion
           '-d', str(doctree_dir), str(source_dir), str(html_dir), *extra_args]
```

File: lib/matplotlib/tests/test_sphinxext.py [hunk 4]
```diff
@@ -97,10 +97,33 @@ def plot_directive_file(num):
     assert filecmp.cmp(range_6, plot_file(5))
 
 
-def build_sphinx_html(source_dir, doctree_dir, html_dir):
+def test_plot_html_show_source_link(tmpdir):
+    source_dir = Path(tmpdir) / 'src'
+    source_dir.mkdir()
+    shutil.copyfile(Path(__file__).parent / 'tinypages' / "conf.py", source_dir / "conf.py")
+    shutil.copytree(Path(__file__).parent / 'tinypages' / "_static", source_dir / "_static")
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_sphinxext.py:104]: Can combine this, but at the very least, use consistent quotes on a line.
```suggestion
    shutil.copyfile(Path(__file__).parent / 'tinypages/conf.py', source_dir)
    shutil.copytree(Path(__file__).parent / 'tinypages/_static', source_dir)
```

File: lib/matplotlib/tests/test_sphinxext.py (hunk 3 above)
Comment (harupy) [reply lib/matplotlib/tests/test_sphinxext.py:126]: @QuLogic Thanks for the suggestions!

File: lib/matplotlib/tests/test_sphinxext.py [hunk 5]
```diff
@@ -97,10 +97,33 @@ def plot_directive_file(num):
     assert filecmp.cmp(range_6, plot_file(5))
 
 
-def build_sphinx_html(source_dir, doctree_dir, html_dir):
+def test_plot_html_show_source_link(tmpdir):
+    source_dir = Path(tmpdir) / 'src'
+    source_dir.mkdir()
+    shutil.copyfile(Path(__file__).parent / 'tinypages/conf.py', source_dir)
```
Comment (harupy) [inline lib/matplotlib/tests/test_sphinxext.py:103]: https://github.com/matplotlib/matplotlib/pull/20814/checks?check_run_id=3318293981#step:12:185

```
E           IsADirectoryError: [Errno 21] Is a directory: '/tmp/pytest-of-runner/pytest-0/popen-gw1/test_plot_html_show_source_lin0/src'
```

It seems that the destination must be a file.

Comment (harupy) [reply lib/matplotlib/tests/test_sphinxext.py:103]: I'll replace `shutil.copyfile` with `shutil.copy`.

Comment (QuLogic) [reply lib/matplotlib/tests/test_sphinxext.py]: Oops, sorry about that.

Comment (harupy) [reply lib/matplotlib/tests/test_sphinxext.py]: No problem!

**Comment by harupy** [comment]
@QuLogic Could you take another review?

**Review by QuLogic — APPROVED** [review]

**Comment by QuLogic** [comment]
Thanks @harupy! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

**Pull request merged**